### PR TITLE
scoped routes: Add ability to derive key fragments from dynamic metadata

### PR DESCRIPTION
**Pull request by tonya11en**
This patch adds support for scope key derivation from dynamic metadata. Prior to this, the scope_key_builder's [fragments](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto#envoy-v3-api-msg-extensions-filters-network-http-connection-manager-v3-scopedroutes-scopekeybuilder-fragmentbuilder) only supported extracting values from headers. With the dynamic metadata keys, we can now derive scope from things like connection metadata as long as it's included in the metadata map.

**Risk Level:**
Low. The change is additive and makes minor changes to existing code paths.
**Testing:**
New unit tests.

**Docs Changes:**
Scoped route docs have a new example with the metadata derivation.

**Release Notes:**
Updated

Fixes https://github.com/envoyproxy/envoy/issues/16945

### CODE REVIEW

File: source/common/router/scoped_config_impl.cc [hunk 1]
```diff
@@ -16,65 +18,107 @@ bool ScopeKey::operator==(const ScopeKey& other) const {
...
+  }
+}
+
+void FragmentBuilderImpl::validateMetadataValueExtractorConfig(
+    const MetadataValueExtractorConfig& config) const {
+  // Assuming support for only a single type of metadata extraction.
+  ASSERT(config.metadata_extract_type_case() == MetadataValueExtractorConfig::kMetadataKey);
+}
+
+FragmentBuilderImpl::FragmentBuilderImpl(ScopedRoutes::ScopeKeyBuilder::FragmentBuilder config)
```
Comment (rojkov) [inline source/common/router/scoped_config_impl.cc:40]: A drive-by question: why is `config`  not rvalue now? This ctor seems to be called only once with `ScopedRoutes::ScopeKeyBuilder::FragmentBuilder` instantiated at the call site.

Comment (tonya11en) [reply source/common/router/scoped_config_impl.cc:40]: You're right, but it's still an rvalue. It's just not forced on the callers.

File: api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto [hunk 2]
```diff
@@ -894,11 +895,24 @@ message ScopedRoutes {
...
 
+      // Specifies how dynamic metadata values should be extracted.
+      message DynamicMetadataValueExtractor {
+        oneof metadata_extract_type {
+          option (validate.required) = true;
+
+          // Extract the entire string value for the provided metadata key.
+          type.metadata.v3.MetadataKey metadata_key = 1;
+        }
+      }
```
Comment (mattklein123) [inline api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto:906]: What other options do you foresee here? Trying to understand how extensible we want this section of the API to be.

File: docs/root/intro/arch_overview/http/http_routing.rst [hunk 3]
```diff
@@ -75,6 +75,21 @@ If the "addr" header value is "foo=1;x-foo-key=127.0.0.1;x-bar-key=1.1.1.1", the
           key: x-foo-key
           separator: =
 
+One may also choose derive scope keys from dynamic :ref:`metadata
```
Comment (mattklein123) [inline docs/root/intro/arch_overview/http/http_routing.rst:78]:
```suggestion
One may also choose to derive scope keys from dynamic :ref:`metadata
```

?

File: envoy/router/scopes.h [hunk 4]
```diff
@@ -92,14 +92,19 @@ class ScopedConfig : public Envoy::Config::ConfigProvider::Config {
    * @param headers the request headers to match the scoped routing configuration against.
    * @return ConfigConstSharedPtr the router's Config matching the request headers.
    */
-  virtual ConfigConstSharedPtr getRouteConfig(const Http::HeaderMap& headers) const PURE;
+  virtual ConfigConstSharedPtr
+  getRouteConfig(const Http::HeaderMap& headers,
```
Comment (mattklein123) [inline envoy/router/scopes.h:96]: nit: update params here and below.

File: source/common/http/conn_manager_impl.cc [hunk 5]
```diff
@@ -551,8 +551,9 @@ void ConnectionManagerImpl::RdsRouteConfigUpdateRequester::requestRouteConfigUpd
     requestVhdsUpdate(host_header, thread_local_dispatcher, std::move(route_config_updated_cb));
     return;
   } else if (parent_.snapped_scoped_routes_config_ != nullptr) {
-    Router::ScopeKeyPtr scope_key =
-        parent_.snapped_scoped_routes_config_->computeScopeKey(*parent_.request_headers_);
+    const auto& dynamic_metadata = parent_.connection()->streamInfo().dynamicMetadata();
```
Comment (mattklein123) [inline source/common/http/conn_manager_impl.cc:554]: It has bitten us more times than I would like that we have different stream info: in this case we have both connection and stream level stream info. Is there ever going to be a case where we want to actually pull metadata from an HTTP filter? I'm not super familiar with this code, but for example, couldn't there be a filter that runs before the on_demand filter which could set some metadata? Do we want to make this more clear in the docs and potentially also prepare the API for supporting stream level stream info vs. connection level?

File: source/common/router/scoped_config_impl.cc [hunk 6]
```diff
@@ -16,65 +18,108 @@ bool ScopeKey::operator==(const ScopeKey& other) const {
...
+  switch (config_.type_case()) {
+  case FragmentBuilderConfig::kHeaderValueExtractor:
+    validateHeaderValueExtractorConfig(config_.header_value_extractor());
+    break;
+  case FragmentBuilderConfig::kMetadataValueExtractor:
+    validateMetadataValueExtractorConfig(config_.metadata_value_extractor());
+    break;
+  case FragmentBuilderConfig::TYPE_NOT_SET:
+    ASSERT(false, "value extractor is not set.");
+    throw ProtoValidationException("value extractor is not set.", config);
```
Comment (mattklein123) [inline source/common/router/scoped_config_impl.cc:51]: I don't think this can happen due to proto validation? And below should be NOT_REACHED?

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Thanks makes sense to me at a high level. A few questions/comments to get started.

/wait

File: source/common/http/conn_manager_impl.cc (hunk 5 above)
Comment (tonya11en) [reply source/common/http/conn_manager_impl.cc:554]: This actually may have just bitten me. I didn't realize that this `StreamInfo` inside the connection was specific to the connection and information set in the HTTP filters won't show up here. I actually need dynamic metadata set in an HTTP filter for my use-case.

Would it make sense to have a distinct `metadata_extract_type` for connections and streams?  In the meantime, I'll try and make some changes to the API to allow us to specify which exact metadata type we want to look up. 

I think I'm going to also beef up those docs around this to make it more clear. It wasn't obvious to me at all that this is the case.

**Comment by tonya11en** [comment]
Keeping this one fresh. Working on other, staler PRs and will get to this soon.

**Comment by mattklein123** [comment]
Is this ready to review? Can you merge main and I will take a look if so?

/wait

**Comment by tonya11en** [comment]
@mattklein123 this is ready for review.

File: source/common/http/conn_manager_impl.cc [hunk 7]
```diff
@@ -565,8 +565,11 @@ void ConnectionManagerImpl::RdsRouteConfigUpdateRequester::requestRouteConfigUpd
     requestVhdsUpdate(host_header, thread_local_dispatcher, std::move(route_config_updated_cb));
     return;
   } else if (parent_.snapped_scoped_routes_config_ != nullptr) {
-    Router::ScopeKeyPtr scope_key =
-        parent_.snapped_scoped_routes_config_->computeScopeKey(*parent_.request_headers_);
+    // @tallen
```
Comment (mattklein123) [inline source/common/http/conn_manager_impl.cc:568]: ?

File: source/common/router/scoped_config_impl.cc [hunk 8]
```diff
@@ -16,65 +18,112 @@ bool ScopeKey::operator==(const ScopeKey& other) const {
...
 
+std::unique_ptr<ScopeKeyFragmentBase>
+FragmentBuilderImpl::computeFragmentFromMetadata(const Metadata& meta,
+                                                 const MetadataValueExtractorConfig& config) const {
+  const auto& value = Envoy::Config::Metadata::metadataValue(&meta, config.metadata_key());
+  switch (value.kind_case()) {
+  case ProtobufWkt::Value::kStringValue:
+    return std::make_unique<StringKeyFragment>(value.string_value());
+  default:
+    // TODO (tonya11en): Support structured metadata fragments.
```
Comment (mattklein123) [inline source/common/router/scoped_config_impl.cc:122]: Sorry what does this mean? Do we need to raise this as an error/warning/etc.? Or at least a debug log?

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Thanks generally LGTM. IIRC we have SRDS integration tests. Is it possible to add an integration test for this? It should'd be too hard with test filters I think.

/wait

File: source/common/router/scoped_config_impl.cc (hunk 8 above)
Comment (tonya11en) [reply source/common/router/scoped_config_impl.cc:122]: Folks can potentially pack [typed metadata](https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/advanced/data_sharing_between_filters.html#typed-metadata) in the map (like kList or kStruct types) as opposed to just string values. There's potential here for more complicated matching logic based on these more complex/custom types.

**Comment by tonya11en** [comment]
busy with other PRs at the moment, I'll get to this after

**Comment by tonya11en** [comment]
I had to put it on the back-burner as other work took precedence. I'll be able to jump back on this after EnvoyCon. We still need this functionality.

**Comment by tony612** [comment]
What's the status of this PR?

**Comment by tony612** [comment]
@tonya11en Any plan to launch this?
